features/323-login-id/type.md

```diff
@@ -0,0 +1,94 @@
+# Login ID Type
+
+## Background
+
+Skygear Auth supports login with login id and password. This document defines the types of login id.
+
+## Validation and Normalization
+
+Validation and normalization are defined base on login id type. When user signup with login ids, Skygear Auth will validate the login ids and reject the signup if it won't pass. After passing the validation, Skygear Auth will store the normalized login id for login.
+
+When user login with login id, login id key is optional. If login id key is provided, Skygear Auth will normalize the input by type, query the principal and perform the login. If login id key is not provided, Skygear Auth will walk through all login id keys, normalize the input and query the principal. If multiple principals are found, the login process will be stopped and error will be returned.
+
+## Types
+
+This section defines the login id types and its validation and normalization rules.
+
+- [Email](#Email)
+- [Username](#Username)
+- [Phone](#Phone)
+- [Raw](#Raw)
+
+### Email
+
+#### Validation
+
+- [RFC 5322 address](https://tools.ietf.org/html/rfc5322#section-3.4.1) (Compulsory)
```

Comment:
So both the local part and the domain part are ASCII only, right? If it is not the case, then we may also need NFKC and Unicode case folding in normalization.

Reply:
Some additional information: https://en.wikipedia.org/wiki/International_email

Reply:
I don't think local part nor the domain part of an email address should be ASCII only, it needs to support unicode...

ICANN have a concept called [Universal Acceptance](https://uasg.tech/wp-content/uploads/documents/UASG004-en-digital.pdf), which is to make sure all software support email address and domain name correctly....

To quickly grab the concept, can look at their [list of test cases](https://uasg.tech/wp-content/uploads/documents/UASG004-en-digital.pdf) of domain name and email address a system should support.



ICANN is promoting a concept called Universal Acceptance.

Reply:
Ok, let us change to perform NFKC to all email address, always case fold domain part and case fold the local part if case sensitive setting is Off. And local part default case sensitive setting is Off.

Reply:
> always case fold domain part

We need to encode the domain properly: https://godoc.org/golang.org/x/net/idna

Reply:
Do you mean conversing the domain to punycode? @louischan-oursky mentioned this too, but we are not sure should we store the unicode version, and the email service should be able to handle the unicode domain?

Reply:
At least we should ensure that a punycode-encoded email should be treated as duplication of raw email.

Reply:
ummm... maybe we should convert it. We will want to avoid user signup 2 accounts that actually the same email, one is unicode version and another is ascii code version

Reply:
@kiootic What do you meant by encode? If you meant encode to punycode, I think we would need some special handling.. it is OKAY to match login ID with punycode, but we would probably want to display it as "normal" most of the time...

Alternatively, is it good if we just use the [ValidateForRegistration()](https://godoc.org/golang.org/x/net/idna#ValidateForRegistration) to validate the email address part? and don't do any "normalization" on the domain part?

BTW, If we want to be really good at displaying the domain name securely, can reference [how Firefox handle it](https://wiki.mozilla.org/IDN_Display_Algorithm) -- but I think this is probably not worth the time now and should be another feature issue...

Reply:
@kiootic @carmenlau a tricky thing is if we store domain part in punycode, we probably need to make it easy for users to display the unicode version of the loginID.

Any idea how to do it? Or we left it for developers? @@

Another thing is, if we convert to punycode, do we still need to call `ValidateForRegistration()` to validate the email address part?

BTW, If we want to be really good at displaying the domain name securely, can reference [how Firefox handle it](https://wiki.mozilla.org/IDN_Display_Algorithm) -- but I think this is probably not worth the time now and should be another feature issue (create a feature issue for it? @carmenlau )